src/App.js

```diff
@@ -1,36 +1,49 @@
-import React, { Component } from 'react';
+import React from 'react';
 import './App.css';
-import { Web3Provider } from 'react-web3';
 import AddTokenPanel from './AddTokenPanel';
 import EditTokenPanel from './EditTokenPanel';
 import DownloadMetaMaskButton from './DownloadMetaMaskButton';
 import { BrowserRouter, Route, Switch, HashRouter } from 'react-router-dom';
 
-class App extends Component {
-
-  render(props, context) {
-    return (
-      <div className="App">
-        <a className="github-banner" href="https://github.com/MetaMask/Add-Token"><img src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub" /></a>
-        <Web3Provider
-          web3UnavailableScreen={() => <div>
-            <p>You need a web3 browser like MetaMask to use this site and manage cryptocurrencies.</p>
-            <DownloadMetaMaskButton/>
-          </div>}
-        >
-          <BrowserRouter basename={process.env.PUBLIC_URL}>
-            <HashRouter hashType="noslash">
-              <Switch>
-                <Route path="/edit" component={EditTokenPanel} />
-                <Route path="/add" component={AddTokenPanel} />
-                <Route path="/" component={AddTokenPanel} />
-              </Switch>
-            </HashRouter>
-          </BrowserRouter>
-        </Web3Provider>
-      </div>
-    );
-  }
+
+const MainContent = () => {
+  return (
+    <div>
+      <BrowserRouter basename={process.env.PUBLIC_URL}>
+        <HashRouter hashType="noslash">
+          <Switch>
+            <Route path="/edit" component={EditTokenPanel} />
+            <Route path="/add" component={AddTokenPanel} />
+            <Route path="/" component={AddTokenPanel} />
+          </Switch>
+        </HashRouter>
+      </BrowserRouter>
+    </div>
+  )
+};
+
+const ErrorContent = () => {
+  return (
+    <div>
+      <p>You need a web3 browser like MetaMask to use this site and manage cryptocurrencies.</p>
+      <DownloadMetaMaskButton />
+    </div>
+  )
 }
 
+const App = () => {
+  const { ethereum } = window;
+
+  return (
+    <div className="App">
+      <a className="github-banner" href="https://github.com/MetaMask/Add-Token"><img src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub" /></a>
+
+      {ethereum ? <MainContent /> : <ErrorContent />}
```

Comment:
This might not work correctly in cases where `window.ethereum` is injected asynchronously, like with MetaMask mobile. On pageload `window.ethereum` can be `undefined`, and React wouldn't know to re-render when it was set. We might need something more sophisticated here.

I'd suggest maybe adding a third state for loading, and calling `detectEthereumProvider` in a `useEffect` hook so that you can get the provider upon mount.

You could also consider setting the provider in React context, so that you can use it directly in the other components rather than calling `detectEthereumProvider` separately in each case. Not necessary though, what you have in the other components right now works.